---
title: "Report"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preprocessing

```{mapping}
# module load cutadapt; module load trimgalore; ls *fastq.gz | awk '{split($0,a,"_"); print a[3]}' | uniq | parallel --eta -j1 trim_galore --paired SO_9588_{}_R1.fastq.gz SO_9588_{}_R2.fastq.gz --fastqc > trim_galore.log

# module load STAR/2.7.3a; STAR --runThreadN 15 --runMode genomeGenerate --genomeDir  ~/third_party_sofware/hg38/karyotypic/ --genomeFastaFiles ~/third_party_sofware/hg38/karyotypic/Homo_sapiens_assembly38.fasta --sjdbGTFfile ~/third_party_sofware/hg38/karyotypic/hg38_ucsc.gtf --sjdbOverhang 100

# module load STAR/2.7.3a; for i in $(ls *_val_*fq.gz | rev | cut -c 16- | rev | uniq); do mkdir -p alignment_hg38/$i; cd alignment_hg38/$i; STAR --genomeDir ~/third_party_sofware/hg38/karyotypic/ --readFilesIn ../../$i'_R1_val_1.fq.gz' ../../$i'_R2_val_2.fq.gz' --readFilesCommand zcat --genomeLoad LoadAndKeep --runThreadN 42 --outSAMtype BAM SortedByCoordinate --limitBAMsortRAM 40000000000 --outReadsUnmapped Fastx; cd ../../; done
```

Import libraries

```{r libraries, message=FALSE, warning=FALSE, echo=FALSE}
library(tidyverse)
library(tidySummarizedExperiment)
library(tidybulk)
library(tidyHeatmap)
library(DT)
library(patchwork)
```

## Gene counting and Data preprocessing

Include filtering, scaling, dimensionality reduction

```{r counting}
counts_scaled = 
  read_csv("company_derived_normalised_counts.csv") %>%
  pivot_longer(cols = -Gene_Name, names_to = "sample", values_to = "count") %>%
  
  # Add factor of interest
  inner_join(
    tibble(
      sample = c("W1", "W2", "W3", "Ko1", "Ko2", "Ko3"),
      type = c(rep("Wild_type", 3), rep("Knock_out", 3))
    ),
    by="sample"
  ) %>%
  as_SummarizedExperiment(.sample = sample, .transcript = Gene_Name , .abundance = count) %>%
  
  # Filter
  identify_abundant(factor_of_interest=type) %>%
  
  # Scale
  scale_abundance() %>%
  
  # Reduce dimensions
  reduce_dimensions(method = "PCA")
```


## Density plot

The libraries are correctly scaled and have quite low depth

```{r density}
counts_scaled %>%
  keep_abundant(factor_of_interest=type) %>%
  ggplot(aes(count_scaled+1, group=.sample, color=type)) +
  geom_density() +
  scale_x_log10() +
  scale_color_brewer(palette = "Set1") +
  theme_bw()
```

## Principal component analysis

The biological groups are quite separated, except for Ko1


```{r PCA}
counts_scaled %>%
  as_tibble() %>%
  nanny::subset(.sample) %>%
  ggplot(aes(PC1,PC2, label=.sample)) +
  geom_point(aes(color=type)) +
  ggrepel::geom_text_repel() +
  scale_color_brewer(palette = "Set1") +
  theme_bw()

ggsave(
  "PCA.pdf",
  useDingbats=FALSE,
  units = c("mm"),
  width = 183 ,
  height = 183,
  limitsize = FALSE
)
```

# Heatmap

The most variable genes put Ko1 more similar to wild-type

```{r heatmap}
plot_heatmap = 
  counts_scaled %>%
  keep_variable(.abundance = counts_scaled, top = 500) %>%
  as_tibble() %>%
  heatmap(
    .column = .sample,
    .row = .feature,
    .value = count_scaled,
    transform = log1p
  ) %>%
  add_tile(type, palette  = c("#E41A1C" ,"#377EB8")) 

plot_heatmap

plot_heatmap %>% save_pdf("heatmap_variable_genes.pdf")

```


## Differential gene feature-abundance

Robust and stringent analyses show a total of `522` downregulated genes and `468` up-regulated genes with a `2-fold` difference between wild-type and knock-out

```{r DE}
counts_de = 
  counts_scaled %>%
  test_differential_abundance(
    ~0+type, 
    method="edger_robust_likelihood_ratio", 
    .contrasts = "typeKnock_out-typeWild_type", 
    test_above_log2_fold_change = 1,
    omit_contrast_in_colnames=TRUE
  )
saveRDS(counts_de, "counts_de.rds")

counts_de %>%
  as_tibble() %>%
  nanny::subset(.feature) %>%
  arrange(FDR) %>%
  describe_transcript(.feature) %>%
  dplyr::select(.feature, description, logFC, FDR) %>%
  DT::datatable()
```



## Gene .feature-abundance distribution

## Heatmap with significant genes

These genes define the two biological groups as expected

```{r heatmap DE}
# Heatmap
plot_heatmap_de = 
  counts_de %>%
  filter(FDR<0.05)%>% #  & abs(logFC) >=2
  as_tibble() %>%
  heatmap(
    .column = .sample,
    .row = .feature,
    .value = count_scaled,
    transform = log1p
  ) %>%
  add_tile(type, palette  = c("#E41A1C" ,"#377EB8")) 
#%>%
#   InteractiveComplexHeatmap::ht_shiny()

plot_heatmap_de

plot_heatmap_de %>% save_pdf("heatmap_differential_genes.pdf", width = 183, height = 183, units = "mm")

```

## Volcano plot

```{r}
topgenes_symbols <-
  counts_de %>%
  as_tibble() %>%
  nanny::subset(.feature)  %>%
  filter(FDR<0.05) %>%
  arrange(desc(abs(logFC)))  %>%
  pull(.feature)

counts_de %>%
  as_tibble() %>%
  nanny::subset(.feature) %>%
  
  # Subset data
  mutate(significant = case_when(
    FDR<0.05 & logFC < 0 ~ "Down-regulated", 
    FDR<0.05 & logFC > 0 ~ "Up-regulated",
    TRUE ~ ""
  )) %>%
  mutate(.feature = ifelse(.feature %in% topgenes_symbols, as.character(.feature), "")) %>%
  
  # Plot
  ggplot(aes(x = logFC, y = PValue, label=.feature)) +
  geom_vline(xintercept = c(-1, 1), linetype="dashed", color="#8c8c8c") +
  geom_hline(yintercept = mean(c(1.24E-4, 3.74E-4)), linetype="dashed", color="#8c8c8c") +
  geom_point(aes(color = significant,  alpha=significant)) +
  ggrepel::geom_text_repel() +
  
  
  # Custom scales
  xlab("Log 2 fold change") +
  theme_bw() +
  scale_y_continuous(trans = "log10_reverse") +
  scale_color_manual(values=c("black", "#1c4966", "#e11f28")) +
  scale_size_discrete(range = c(0, 2))

ggsave(
  "volcano_plot.pdf",
  useDingbats=FALSE,
  units = c("mm"),
  width = 183 ,
  height = 183,
  limitsize = FALSE
)
```

## EGSEA gene enrichment

EGSEA uses many algorithms on many datasets (GSEA, GO, KEGG). The pathways are ranked by the overall gene set that scores well in most algorithms. The web_page column includes the web page where ll information of that gene set are.

Please download from the email the compressed directory, and open index.html with a browser to investigate the results for each pathway in the table below.

```{r EGSEA, eval=FALSE}

counts_gene_rank = 
  counts_de %>%
  symbol_to_entrez(.feature) %>%
  filter(PValue   %>% is.na %>% `!`) %>%
  test_gene_rank(
    .entrez = entrez,
    .arrange_desc = logFC ,
    species="Homo sapiens",
    gene_sets = c("H", "C2", "C5")
  )
```

```{r load 2, echo=FALSE}
counts_gene_rank = readRDS("counts_gene_rank.rds")

```

Curated aging gene sets, look p.adjust < 0.05

```{r}

aging_sets = 
  counts_gene_rank %>%
  filter(gs_cat  == "C2" ) %>%
  unnest(test) %>%
  rowid_to_column(var = "GSEA rank") %>%
  filter(grepl("aging", ID, ignore.case = T)) %>%
  filter(p.adjust < 0.05) 

aging_sets %>%
    dplyr::select(-fit) %>%
  DT::datatable()

```

```{r}
aging_sets %>%
  mutate(aging = if_else(grepl("UP", ID), "Aging up", "Aging down")) %>%
  ggplot(aes(forcats::fct_reorder(ID, enrichmentScore), enrichmentScore, fill=p.adjust, size=Count)) +
  #ggplot(aes(forcats::fct_reorder(ID, enrichmentScore), enrichmentScore,size=Count, color=p.adjust )) +
  geom_hline(yintercept = 0, color="#c8c8c8", linetype="dashed") +
  geom_point(shape = 21) +
  facet_grid(~ aging , scales = "free", space = "free") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  scale_fill_distiller( trans = "log10_reverse", palette = "Spectral") +
  #scale_y_discrete(labels = label_func) +
  ylab("Enrichment score") + 
  xlab(NULL) +
  scale_size(range=c(3, 8))


ggsave(
  "aging_sets_point.pdf",
  useDingbats=FALSE,
  units = c("mm"),
  width = 183/2 ,
  height = 183/2,
  limitsize = FALSE
)
```


```{r}
aging_sets %>% 
  mutate(plot = pmap(
    list(fit, ID, idx_for_plotting, p.adjust), 
    ~ enrichplot::gseaplot2(
      ..1, 
      geneSetID = ..3, 
      title = sprintf("%s \nadj pvalue %s", ..2, round(..4, 2)),
      base_size = 6
    ) 
  )) %>%
  pull(plot)%>%
  wrap_plots()

ggsave(
  "aging_sets_trends.pdf",
  useDingbats=FALSE,
  units = c("mm"),
  width = 183 ,
  height = 183,
  limitsize = FALSE
)
```

All curated gene sets, look p.adjust < 0.05

```{r}

counts_gene_rank %>%
  filter(gs_cat  == "C2" ) %>%
  dplyr::select(-fit) %>%
  unnest(test) %>%
  DT::datatable()

```

All hallmarks, look p.adjust < 0.05

```{r}

counts_gene_rank %>%
  filter(gs_cat  == "H" ) %>%
  dplyr::select(-fit) %>%
  unnest(test) %>%
  DT::datatable()

```


Overall aging signature

```{r}
C2_genes = 
  aging_sets %>% 
  select(ID, core_enrichment) %>% 
  filter(grepl("UP", ID)) %>% 
  separate_rows(core_enrichment, sep="/") %>%
  left_join(
     counts_de %>%
      symbol_to_entrez(.feature) %>% 
       select(.feature, entrez),
     by = c("core_enrichment" = "entrez")
  ) %>% 
  select(.feature, data_base = ID) %>% 
  distinct()


vijay_genes = 
  read_csv("combined_age_gene_signature.csv") %>% 
  select(.feature = X1, data_base) %>% 
  distinct()

aging_genes = 
  C2_genes %>% bind_rows(vijay_genes)

aging_rank = 
  counts_de %>%
       symbol_to_entrez(.feature) %>%
         filter(PValue   %>% is.na %>% `!`) %>%
  test_gene_rank(
    .entrez = entrez,
    .arrange_desc = logFC ,
    species="Homo sapiens",
    gene_sets = list(
      aging_signature = 
        (counts_de %>%
         filter(.feature %in% aging_genes$.feature) %>% 
         symbol_to_entrez(.feature) %>% 
        pull(entrez) %>% 
        as.integer())
    )
  )
```

Plot Heatmap
```{r}
#Heatmap
aging_heatmap = 
  counts_de %>% 
  inner_join(aging_genes) %>% 
  mutate(data_base = tolower(data_base) ) %>% 
  unite(".feature", c(.feature, data_base), remove = F) %>% 
  group_by( data_base) %>% 
   heatmap(
    .column = .sample,
    .row = .feature,
    .value = count_scaled,
    transform = log1p,
    palette_value = circlize::colorRamp2(
            seq(-2, 2, length.out = 11), 
            RColorBrewer::brewer.pal(11, "RdBu")
        )
  ) %>% 
  add_tile(type)


aging_heatmap
```

Plot GSEA 
```{r}

plot_aging_rank =
  aging_rank%>% 
  unnest(test) %>% 
  mutate(plot = pmap(
     list(fit, ID, idx_for_plotting, p.adjust), 
     ~ enrichplot::gseaplot2(
         ..1, 
         geneSetID = ..3, 
         title = sprintf("%s \nadj pvalue %s", ..2, round(..4, 2)),
         base_size = 6
     ) 
 )) %>%
     pull(plot) %>% 
  .[[1]]

plot_aging_rank
```

# Bibliography

```{r}

counts_gene_rank %>% get_bibliography() %>%
  c(counts_gene_rank %>% get_bibliography()) %>%
  unique()
```

# Session information

```{r}
sessionInfo()
```


